test/e2e/apibinding/apibinding_test.go

```diff
@@ -442,29 +442,46 @@ func TestAPIBinding(t *testing.T) {
 	export2, err := kcpClusterClient.Cluster(provider2Path).ApisV1alpha2().APIExports().Get(ctx, exportName, metav1.GetOptions{})
 	require.NoError(t, err)
 
-	foundOnShards := 0
-	//nolint:staticcheck // SA1019 VirtualWorkspaces is deprecated but not removed yet
-	for _, vw := range export2.Status.VirtualWorkspaces {
-		vw2ClusterClient, err := kcpdynamic.NewForConfig(apiexportVWConfig(t, rawConfig, vw.URL))
-		require.NoError(t, err)
+	kcptestinghelpers.Eventually(t, func() (bool, string) {
+		foundOnShards := 0
+		var listErrs []error
+
+		//nolint:staticcheck // SA1019 VirtualWorkspaces is deprecated but not removed yet
+		for _, vw := range export2.Status.VirtualWorkspaces {
+			vw2ClusterClient, err := kcpdynamic.NewForConfig(apiexportVWConfig(t, rawConfig, vw.URL))
+			require.NoError(t, err)
+
+			t.Logf("Listing %s|%s cowboys via virtual workspace %s/clusters/%s", provider2Path, exportName, vw.URL, consumer3ClusterName)
+			gvr := wildwestv1alpha1.SchemeGroupVersion.WithResource("cowboys")
+			list, err := vw2ClusterClient.Cluster(consumer3ClusterName.Path()).Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
+			if err != nil {
+				t.Logf("Error: %v", err)
+				listErrs = append(listErrs, err)
+				continue
+			}
+			require.Equal(t, 1, len(list.Items), "unexpected # of cowboys through virtual workspace with explicit workspace")
+			foundOnShards++
 
-		t.Logf("Listing %s|%s cowboys via virtual workspace %s/clusters/%s", provider2Path, exportName, vw.URL, consumer3ClusterName)
-		gvr := wildwestv1alpha1.SchemeGroupVersion.WithResource("cowboys")
-		list, err := vw2ClusterClient.Cluster(consumer3ClusterName.Path()).Resource(gvr).Namespace("").List(ctx, metav1.ListOptions{})
-		if err != nil {
-			t.Logf("Error: %v", err)
-			continue
+			t.Logf("Listing %s|%s cowboys via virtual workspace wildcard list", provider2Path, exportName)
+			list, err = vw2ClusterClient.Resource(gvr).List(ctx, metav1.ListOptions{})
+			require.NoError(t, err, "error listing through virtual workspace wildcard")
+			require.Equal(t, 1, len(list.Items), "unexpected # of cowboys through virtual workspace with wildcard")
```

Comment:
I suspect all `require` calls (here and one below) should instead be returning `false` to the `Eventually` loop, no? Wouldn't it fail the test case in case of flakes otherwise?

Reply:
Yes that's intentional, with three cases to cover:

* zero elements found: that's the flake, and we want to retry,
* one element: we're good and can return with success,
* more than one: definitely should not happen, and retrial won't help so we fail the test.

Does this sound reasonable?

Reply:
Similarly with the # of `cowboys`, if it's not one then something else must be broken and we fail the test.

Reply:
An fail raised by `require` is collected by `Eventually`: https://github.com/stretchr/testify/blob/v1.10.0/assert/assertions.go#L2016
I had the same question when I worked on flakes but earlier `require` fails are ignored once the condition goes through.

Reply:
I see! TIL. Thanks for educating me.